tools/cli/helpers/doc-parser/src/class-doc-parser.php

```diff
@@ -104,15 +146,304 @@ function ( $item ) use ( $ignore ) {
 			}
 		);
 
+		$nodeFinder = new NodeFinder();
+
 		// Extract PHPDoc.
-		ob_start();
-		$output = \WP_Parser\parse_files( $files, $path );
-		ob_end_clean();
+		$blocks = array();
+
+		foreach ( $files as $file ) {
+			printf( 'Extracting PHPDoc from %1$s.' . PHP_EOL, $file );
+
+			$stmts = $this->parser->parse( file_get_contents( $file ) );
+			if ( empty( $stmts ) ) {
+				continue;
+			}
+
+			// Find all calls to apply_filters or do_action.
+			$hookCalls = $nodeFinder->find(
+				$stmts,
+				function ( Node $node ) {
+
+					if ( ! $node instanceof FuncCall ) {
+						return false;
+					}
+
+					return $node->name->name === 'apply_filters'
+						|| $node->name->name === 'do_action';
+				}
+			);
+
+			$file_blocks = $this->get_filter_calls( $hookCalls );
+
+			$splfile = new \SplFileObject( $file );
+			foreach ( $file_blocks as &$block ) {
+
+				$docblock = array();
+
+				// Lines are zero indexed.
+				$start = $block['line'] - 2;
+
+				$first = true;
+				while ( ! $splfile->eof() && $start >= 0 ) {
+					$splfile->seek( $start-- );
+					$line = $splfile->current();
+
+					if ( $first && false === strpos( $line, '*/' ) ) {
+
+						break;
+					} else {
+						$first = false;
+					}
+
+					array_unshift( $docblock, $line );
+					if ( false !== strpos( $line, '/**' ) ) {
+						break;
+					}
+				}
+				$docblock = implode( '', $docblock );
+
+				$block['doc']                     = array();
+				$block['doc']['description']      = '';
+				$block['doc']['long_description'] = '';
+				$block['doc']['tags']             = array();
+
+				try {
+					$tokens     = new TokenIterator( $this->lexer->tokenize( $docblock ) );
+					$phpDocNode = $this->pdparser->parse( $tokens );
+				} catch ( ParserException $e ) {
+					continue;
+				}
+
+				foreach ( $phpDocNode->children as $entry ) {
+					if ( ! $entry instanceof PhpDocTextNode ) {
+						continue;
+					}
+
+					if ( ! empty( $entry->text ) ) {
+						$block['doc']['description'] .=
+							'<p>'
+							. str_replace( array( "\r\n", "\n", "\r" ), '</p><p>', $entry->text )
+							. '</p>';
+					}
+				}
+
+				$paramTags  = $this->get_param_tag_nodes( $phpDocNode );
+				$parameters = array();
+
+				foreach ( $paramTags as $paramTag ) {
+					$block['doc']['tags'][] = array(
+						'name'     => 'param',
+						'content'  => $paramTag->description,
+						'types'    => array(
+							(string) $paramTag->type,
+						),
+						'variable' => $paramTag->parameterName,
+					);
+
+					$parameters[] = (string) $paramTag . PHP_EOL;
+				}
+
+				foreach (
+					array(
+						'@since',
+						'@module',
+						'@deprecated',
+						'@see',
+						'@uses',
+						'@link',
+						'@type',
+					) as $tagType
+				) {
```

Comment:
Another useful diagnostic would be to complain if any tags other than these are found.